main.tf

```diff
@@ -139,7 +139,7 @@ resource "aws_route" "public" {
 resource "aws_route_table_association" "public" {
   count          = local.public_subnets
   subnet_id      = aws_subnet.public[count.index].id
-  route_table_id = aws_route_table.public[0].id
+  route_table_id = var.shared_public_route_table ? element(aws_route_table.public[*].id, count.index) : aws_route_table.public[0].id
```

Comment:
This still needs to be inverted and I don't think we need to use the `element` function here.